src/Common/src/System/Security/Cryptography/PasswordBasedEncryption.cs

```diff
@@ -0,0 +1,1032 @@
+                        {
+                            CryptographicOperations.ZeroMemory(sourceRent.AsSpan(0, source.Length));
+                            ArrayPool.Return(sourceRent);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static unsafe int Pbes2Decrypt(
+            ReadOnlyMemory<byte>? algorithmParameters,
+            ReadOnlySpan<char> password,
+            ReadOnlySpan<byte> passwordBytes,
+            ReadOnlySpan<byte> encryptedData,
+            Span<byte> destination)
+        {
+            Span<byte> buf = stackalloc byte[128];
+            ReadOnlySpan<byte> effectivePasswordBytes = stackalloc byte[0];
+            byte[] rented = null;
+            System.Text.Encoding encoding = null;
+
+            if (passwordBytes.Length > 0 || password.Length == 0)
+            {
+                effectivePasswordBytes = passwordBytes;
+            }
+            else
+            {
+                encoding = System.Text.Encoding.UTF8;
+                int byteCount = encoding.GetByteCount(password);
+
+                if (byteCount > buf.Length)
+                {
+                    rented = ArrayPool.Rent(byteCount);
+                    buf = rented.AsSpan(0, byteCount);
+                }
+                else
+                {
+                    buf = buf.Slice(0, byteCount);
+                }
+            }
+
+            fixed (byte* maybeRentedPtr = &MemoryMarshal.GetReference(buf))
+            {
+                if (encoding != null)
+                {
+                    int written = encoding.GetBytes(password, buf);
+                    Debug.Assert(written == buf.Length);
+                    buf = buf.Slice(0, written);
+                    effectivePasswordBytes = buf;
+                }
+
+                return Pbes2Decrypt(
+                    algorithmParameters,
+                    effectivePasswordBytes,
+                    encryptedData,
+                    destination);
+            }
+        }
+
+        private static unsafe int Pbes2Decrypt(
+            ReadOnlyMemory<byte>? algorithmParameters,
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> encryptedData,
+            Span<byte> destination)
+        {
+            if (!algorithmParameters.HasValue)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            PBES2Params pbes2Params =
+                AsnSerializer.Deserialize<PBES2Params>(algorithmParameters.Value, AsnEncodingRules.BER);
+
+            if (pbes2Params.KeyDerivationFunc.Algorithm.Value != Oids.Pbkdf2)
+            {
+                throw new CryptographicException(
+                    SR.Format(
+                        SR.Cryptography_UnknownAlgorithmIdentifier,
+                        pbes2Params.EncryptionScheme.Algorithm.Value));
+            }
+
+            Rfc2898DeriveBytes pbkdf2 =
```

Comment:
any reason to put this outside of using? (same question for cipher below

Reply:
Mainly just line length